keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,362 @@
+```
+
+Older servers and clients that were built before the introduction of full dual stack will only be aware of and make use of the original, singular "PodIP" field. It is therefore considered to be the default IP address for the pod. This choice of which pod IP to choose as the default IP address is discussed in the following subsection.
+
+##### Default Pod IP Selection
+When the internal representation of the PodStatus (see the next section) is converted to the "v1 core" versioned API object (see the [Operational Overview section](https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#operational-overview) of the API changes guidelines), a choice needs to be made as to which IP address from the "PodIPs" slice needs to be populated into the singular "PodIP" field. As described earlier, older servers and clients will only be aware of this singular field, so it is considered to be the default IP address for the pod. The following algorithm will be used to select a default pod IP address:
+- Use the first IPv4 address found in the "ips" list returned from the CNI plugin, if any.
+- If there are no IPv4 addresses, then use the first IPv6 address found in the "ips" list.
+
+The IPv4 addresses take precedence here because:
+- Less ambiguity: There should typically be only one IPv4 address assigned to a pod.
+- Consistency with current node IP selection, where IPv4 addresses on the host take precedence.
+
+It's not expected that we'll need any configuration "knobs" to allow users to tweak this algorithm: Any users who care about the choice of default IP address (i.e. anyone who is using older servers and clients in their cluster) should simply configure their CNI plugins to only use a single IP family (IPv4-only or IPv6-only). With this configuration, there should only be one pod IP address to choose as the default.
+
+#### PodStatus Internal Representation:
+The PodStatus internal representation will be modified to use a slice of IPs ("PodIPs") rather than a singular IP ("PodIP"):
+```
+    // IP addresses allocated to the pod. Routable at least within the cluster.
+    // Empty if not yet allocated.
+    // +optional
+    PodIPs []string `json:"podIP,omitempty" protobuf:"bytes,6,opt,name=podIP"`
+```
+This "PodIPs" slice representation should eventually become part of a versioned API (after a period of deprecation for the current versioned APIs).
+
+#### kubelet Startup Configuration for Multiple Pod CIDRs
+A new, plural "pod-cidrs" option for the [kubelet startup configuration](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/) is proposed, in addition to retaining the existing, singular "pod-cidr" option (for backwards compatibility):
+```
+  --pod-cidr   string        [Singular IP CIDR,  Default: ""]
+  --pod-cidrs  stringSlice   [Multiple IP CIDRs, comma separated list of CIDRs, Default: []]
+```
+The singular --pod-cidr argument will become deprecated.
```

Comment:
will these flags be mutually exclusive?  Seems like they should be.  what happens if both are specified (hint: older has to win)